src/community/jdbcstore/src/main/java/org/geoserver/jdbcstore/JDBCResourceStore.java

```diff
@@ -0,0 +1,375 @@
+            return oldResourceStore.get(path);
+        }
+        return new JDBCResource(dir.createEntry(path));
+    }
+        
+    @Override
+    public boolean remove(String path) {
+        return dir.createEntry(path).delete();
+    }
+    
+    @Override
+    public boolean move(String path, String target) {
+        return dir.createEntry(path).renameTo(dir.createEntry(target));
+    }
+    
+    @Override
+    public String toString() {
+        return "JDBCResourceStore";
+    }
+
+    /**
+     * Direct implementation of Resource.
+     * 
+     */    
+    protected class JDBCResource implements Resource {
+        
+        private static final long serialVersionUID = 1829574259576354635L;
+        
+        private final JDBCDirectoryStructure.Entry entry;
+        
+        public JDBCResource(JDBCDirectoryStructure.Entry entry) {
+            assert(entry != null);
+            this.entry = entry;
+        }
+        
+        @Override
+        public String path() {
+            return entry.toString();
+        }
+
+        @Override
+        public String name() {
+            return entry.getName();
+        }
+        
+        private InputStream getIStream() {
+            return entry.getContent();            
+        }
+                
+        @Override
+        public InputStream in() {
+            final Lock lock = lock();
+            try {
+                entry.createResource();
+                return getIStream();
+            } finally {
+                lock.release();
+            }
+        }
+
+        @Override
+        public OutputStream out() {
+            List<ResourceNotification.Event> events = SimpleResourceWatcher.createEvents(this,
+                    ResourceNotification.Kind.ENTRY_CREATE);
+            if (entry.createResource()) {
+                resourceWatcher.changed(new ResourceNotification(path(), ResourceNotification.Kind.ENTRY_CREATE, 
+                        System.currentTimeMillis(), events));
+            }
+            try {
+                return new CachingOutputStreamWrapper(File.createTempFile("out.", entry.getName()));
+            } catch (IOException ex) {
+                throw new IllegalStateException(ex); 
+            }
+        }
+
+        @Override
+        public File file() {
+            if (getType() == Type.DIRECTORY) {
+                throw new IllegalStateException("Directory (not a file)");
+            }
+            final Lock lock = lock();
+            try {
+                return cache.cache(this, false);
+            } catch(IOException ex) {
+                throw new IllegalStateException(ex);
+            } finally {
+                lock.release();
+            }
+        }
+
+        @Override
+        public File dir() {
+            if (getType() == Type.RESOURCE) {
+                throw new IllegalStateException("File (not a directory)");
+            }
+            final Lock lock = lock();
```

Comment:
Can we use try-with-resources for all Lock operations? This will need ```Resource.Lock``` to implement ```AutoCloseable``` (or even ```Closeable```). Perhaps @jodygarnett can comment on this design choice. Backportability?